### ISSUE TITLE
A few point-free opportunities

### ISSUE THREAD

**Issue by subttle** [emilypi/smash#22]
Hi, first of all, thanks for the excellent package!!! I have found it very useful.

The code is certainly great as it is, however, my love for point-free elegance has compelled me to rewrite some of the functions as point-free one-liners; and I thought I'd share just in case you're interested:

```Haskell
-- Data.Wedge
fromWedge ∷ Wedge a b → Maybe (Either a b)
fromWedge = wedge Nothing (Just . Left) (Just . Right)

toWedge ∷ Maybe (Either a b) → Wedge a b
toWedge = maybe Nowhere (either Here There)

quotWedge ∷ Either (Maybe a) (Maybe b) → Wedge a b
quotWedge = either (maybe Nowhere Here) (maybe Nowhere There)
```

```Haskell
-- Data.Smash
toSmash ∷ Maybe (a, b) → Smash a b
toSmash = maybe Nada (uncurry Smash)

fromSmash ∷ Smash a b → Maybe (a, b)
fromSmash = smash Nothing (curry Just)

swapSmash ∷ Smash a b → Smash b a
swapSmash = smash Nada (flip Smash)
```

```Haskell
-- Data.Can
swapCan ∷ Can a b → Can b a
swapCan = can Non Eno One (flip Two)
```

If you prefer the lambda-case style they're in now, or simply don't have the desire to change the code then that is totally okay with me, just thought I'd share in case you liked it. If you would like I could submit a pull request (and I'd substitute out the Unicode syntax to match the rest of the code, of course).

Also, there's a lot more I wrote (just for fun/practice), but left out because I figured the original version would be more preferable, e.g.:

```Haskell
isNada ∷ Smash a b → Bool
isNada Nada = True
isNada _    = False
```

is probably less noisy than:

```Haskell
isNada ∷ Smash a b → Bool
isNada = smash True (const (const False))
```

I've read the Contributing guidelines[1], and I'm not sure if what I have here applies to "case 3" under "Code Quality" but if that is the ruling then by all means please dismiss the ticket and I won't be offended.

Thank you for your time. I hope you have a nice day!

[1] https://github.com/emilypi/smash/blob/bd3ae30bdfef3f92ce5575a1ee10228830afa11b/CONTRIBUTING.md

**Comment by subttle** [comment]
Oh and perhaps it's worth mentioning because I just read in another [open issue](https://github.com/emilypi/smash/issues/9) that you have plans to pull in `Data.These` then you could also potentially add the following:
```Haskell
-- Data.Can
toCan ∷ Maybe (These a b) → Can a b
toCan = maybe Non (these One Eno Two)

fromCan ∷ Can a b → Maybe (These a b)
fromCan = can Nothing (Just . This) (Just . That) (curry (Just . uncurry These))
```
Thanks again for your time!

**Comment by emilypi** [comment]
Hi @subttle 

These look great, and they're certainly valid! The reason I preferred explicit casing to pointfree because it was easier to reason through the logic clearly in my head while writing. At this point, however, I would not be opposed to some of these pointfree improvements existing in the package. 

> is probably less noisy than...

You have good judgement here. I would definitely prefer the latter to the former in the case of `isNada`. I'd welcome a PR.

**Comment by emilypi** [comment]
Merged in #24

**Comment by subttle** [comment]
Hi again, hope I'm not overdoing it but I found just a few more nice candidates in case you're interested:

```Haskell
gatherWedges :: Wedge [a] [b] -> [Wedge a b]
gatherWedges = wedge [] (fmap Here) (fmap There)
```
```Haskell
gatherSmashes :: Smash [a] [b] -> [Smash a b]
gatherSmashes = smash [] (zipWith Smash)
```
```Haskell
gatherCans :: Can [a] [b] -> [Can a b]
gatherCans = can [] (fmap One) (fmap Eno) (zipWith Two)
```
No worries if you'd rather keep it as-is, but happy to make a quick PR if you'd like. Have a nice day!
